### PR TITLE
MSI node changes

### PR DESCRIPTION
**Pull request by rokmicrosoft**

### CODE REVIEW

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 1]
```diff
@@ -67,9 +81,35 @@ export class ApplicationTokenCredentials {
         return this.clientId;
     }
 
-    private getAuthorizationToken(): Q.Promise<string> {
+    private _getMSIAuthorizationToken(): Q.Promise<string> {
         var deferred = Q.defer<string>();
+        let webRequest = new webClient.WebRequest();
+        webRequest.method = "GET";
+        let port = this.msiPort ? this.msiPort : '50342';
```
Comment (vincentdass) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: Can we set this value upfront in constructor?

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 2]
```diff
@@ -49,11 +54,20 @@ export class ApplicationTokenCredentials {
...
         this.activeDirectoryResourceId = activeDirectoryResourceId;
         this.isAzureStackEnvironment = isAzureStackEnvironment;
+        this.scheme = scheme;
+        this.msiPort = msiPort;
     }
 
     public getToken(force?: boolean): Q.Promise<string> {
         if (!this.token_deferred || force) {
-            this.token_deferred = this.getAuthorizationToken();
+            if(this.scheme && this.scheme === "MSI")
```
Comment (vincentdass) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: Can we have the scheme as Enum? (MSI , SPN)

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts (hunk 1 above)
Comment (rokmicrosoft) [reply Tasks/Common/azure-arm-rest/azure-arm-common.ts]: Done.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts (hunk 2 above)
Comment (rokmicrosoft) [reply Tasks/Common/azure-arm-rest/azure-arm-common.ts]: Done.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 3]
```diff
@@ -49,11 +55,20 @@ export class ApplicationTokenCredentials {
         this.authorityUrl = authorityUrl;
         this.activeDirectoryResourceId = activeDirectoryResourceId;
         this.isAzureStackEnvironment = isAzureStackEnvironment;
+        this.scheme = AzureModels.Scheme[scheme];
+        this.msiPort = msiPort ? msiPort : '50342';
```
Comment (kmkumaran) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: This port will no longer be required. Please check the revised mail on this regard.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 4]
```diff
@@ -67,9 +82,36 @@ export class ApplicationTokenCredentials {
         return this.clientId;
     }
 
-    private getAuthorizationToken(): Q.Promise<string> {
+    private _getMSIAuthorizationToken(): Q.Promise<string> {
         var deferred = Q.defer<string>();
+        let webRequest = new webClient.WebRequest();
+        webRequest.method = "GET";
+        webRequest.uri = "http://localhost:"+ this.msiPort + "/oauth2/token?resource="+ this.baseUrl;
```
Comment (kmkumaran) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: This Uri has changed. You will have to react to this.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 5]
```diff
@@ -49,11 +55,20 @@ export class ApplicationTokenCredentials {
         this.authorityUrl = authorityUrl;
         this.activeDirectoryResourceId = activeDirectoryResourceId;
         this.isAzureStackEnvironment = isAzureStackEnvironment;
+        this.scheme = AzureModels.Scheme[scheme];
```
Comment (kmkumaran) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: If there is no scheme, do you want to initialize it as SPN?

File: Tasks/Common/azure-arm-rest/module.json [hunk 6]
```diff
@@ -5,7 +5,8 @@
         "SecretCannotBeEmpty": "secret must be a non empty string.",
         "armUrlCannotBeEmpty": "arm Url must be a non empty string.",
         "authorityUrlCannotBeEmpty": "authority must be a non empty string.",
-        "CouldNotFetchAccessTokenforAzureStatusCode": "Could not fetch access token for azure. Status code: %s, status message: %s",
+        "CouldNotFetchAccessTokenforAzureStatusCode": "Could not fetch access token for Azure. Status code: '%s', status message: %s",
+        "CouldNotFetchAccessTokenforMSIStatusCode": "Could not fetch access token for Managed Service Principal. Please configure Managed Service Identity (MSI) for virtual machine 'https://aka.ms/azure-msi-docs'. Status code: '%s', status message: %s",
```
Comment (kmkumaran) [inline Tasks/Common/azure-arm-rest/module.json]: can you review this message with Roopesh.

File: Tasks/Common/azure-arm-rest/azure-arm-endpoint.ts [hunk 7]
```diff
@@ -28,12 +28,14 @@ export class AzureRMEndpoint {
             this.endpoint = {
                 subscriptionID: tl.getEndpointDataParameter(this._connectedServiceName, 'subscriptionid', true),
                 subscriptionName: tl.getEndpointDataParameter(this._connectedServiceName, 'subscriptionname', true),
-                servicePrincipalClientID: tl.getEndpointAuthorizationParameter(this._connectedServiceName, 'serviceprincipalid', false),
-                servicePrincipalKey: tl.getEndpointAuthorizationParameter(this._connectedServiceName, 'serviceprincipalkey', false),
+                servicePrincipalClientID: tl.getEndpointAuthorizationParameter(this._connectedServiceName, 'serviceprincipalid', true),
```
Comment (kmkumaran) [inline Tasks/Common/azure-arm-rest/azure-arm-endpoint.ts]: Why this change?

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts (hunk 5 above)
Comment (rokmicrosoft) [reply Tasks/Common/azure-arm-rest/azure-arm-common.ts]: It will automatically goes into SPN case if there is no scheme But I will initialize it with SPN.

File: Tasks/Common/azure-arm-rest/azure-arm-endpoint.ts (hunk 7 above)
Comment (rokmicrosoft) [reply Tasks/Common/azure-arm-rest/azure-arm-endpoint.ts]: Initially we have only one case SPN so servicePrincipalClientID and servicePrincipalClientKey is compulsory field But due to introduction of MSI these two are optional.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 8]
```diff
@@ -67,9 +82,55 @@ export class ApplicationTokenCredentials {
...
+            "Metadata": true
+        };
+
+        webClient.sendRequest(webRequest).then(
+            (response: webClient.WebResponse) => {
+                if (response.statusCode == 200) 
+                {
+                    deferred.resolve(response.body.access_token);
+                }
+                else if (response.statusCode == 209 || response.statusCode == 500)
```
Comment (vincentdass) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: should we[ retry for 429](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service#error) instead of 209?

File: Tasks/AzureAppServiceManage/package-lock.json [hunk 9]
```diff
@@ -18,9 +18,9 @@
       "integrity": "sha1-ibTRmasr7kneFk6gK4nORi1xt2c="
     },
     "brace-expansion": {
-      "version": "1.1.8",
-      "resolved": "https://registry.npmjs.org/brace-expansion/-/brace-expansion-1.1.8.tgz",
-      "integrity": "sha1-wHshHHyVLsH479Uad+8NHTmQopI=",
+      "version": "1.1.11",
```
Comment (vincentdass) [inline Tasks/AzureAppServiceManage/package-lock.json]: revert this change.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 10]
```diff
@@ -11,19 +12,24 @@ export class ApplicationTokenCredentials {
...
-        }
+    constructor(clientId: string, domain: string, secret: string, baseUrl: string, authorityUrl: string, activeDirectoryResourceId: string, isAzureStackEnvironment: boolean, scheme?: string, msiClientId?: string) {
 
         if (!Boolean(domain) || typeof domain.valueOf() !== 'string') {
             throw new Error(tl.loc("DomainCannotBeEmpty"));
         }
 
-        if (!Boolean(secret) || typeof secret.valueOf() !== 'string') {
-            throw new Error(tl.loc("SecretCannotBeEmpty"));
+        if((!scheme ||scheme ==='ServicePrincipal')){
```
Comment (vincentdass) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: scheme != 'MSI' ?

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 11]
```diff
@@ -67,9 +82,55 @@ export class ApplicationTokenCredentials {
         return this.clientId;
     }
 
-    private getAuthorizationToken(): Q.Promise<string> {
+    private _getMSIAuthorizationToken(count: number ,timeToWait: number): Q.Promise<string> {
```
Comment (vincentdass) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: nit: rename `count` to `retryCount`

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 12]
```diff
@@ -67,9 +82,55 @@ export class ApplicationTokenCredentials {
...
+            (response: webClient.WebResponse) => {
+                if (response.statusCode == 200) 
+                {
+                    deferred.resolve(response.body.access_token);
+                }
+                else if (response.statusCode == 209 || response.statusCode == 500)
+                {
+                    if(count < retryLimit)
+                    {
+                        let waitedTime = 2000 + timeToWait * 2;
```
Comment (vincentdass) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: Can u explain this?

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts (hunk 10 above)
Comment (rokmicrosoft) [reply Tasks/Common/azure-arm-rest/azure-arm-common.ts]: It looks more promising to check scheme === serviceprincipal and take action instead of scheme != MSI. As in future we need not change this block in case if we add new scheme.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts (hunk 11 above)
Comment (rokmicrosoft) [reply Tasks/Common/azure-arm-rest/azure-arm-common.ts]: DOne. Taken.

**Comment by rokmicrosoft** [comment]
This Pr was merged into [here](https://github.com/Microsoft/vsts-tasks/pull/6981).
